Quickstart javascript express api (#2535)
FusionAuth/fusionauth-site: astro/src/content/quickstarts/quickstart-javascript-express-api.mdx

</Aside>

### Create A Basic {frontmatter.technology} Application

Now you are going to create an {frontmatter.technology} API application. While this section builds a simple API, you can use the same configuration to integrate an existing API with FusionAuth.

We are going to be building an API backend for a banking application called ChangeBank. This API will have two endpoints:
- `make-change`: This endpoint will allow you to call GET with a `total` amount and receive a response indicating how many nickels and pennies are needed to make change. Valid roles are `customer` and `teller`.
- `panic`: Tellers may call this POST endpoint to call the police in case of an incident. The only valid role is `teller`.

Both endpoints will be protected such that a valid JSON web token (JWT) will be required in a cookie in order to be accessed. Additionally, the JWT must have a `roles` claim containing the appropriate role to use the endpoint.

Create the starter app called `mysite` for this API.

```shell
npx express-generator@latest mysite --no-view
cd mysite
npm install

> npm audit fix --force # run this only if npm reports security vulnerabilities

Author: @mark-robustelli - this is the official express generator, and the problem is with its dependencies. Unfortunately this is a common problem for node packages - a workaround could be to commit the code it generates but this would get out of date and then the user wouldn't be able to start 'from scratch', but we'd recommend just leaving in the 'fix' step for now until the dependencies are fixed.

First create the `services` directory under `mysite`.

```shell
mkdir services
```

In this `services` folder, add the file `verifyJWT.js`, and copy in the following code.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-javascript-express-api/main/complete-application/services/verifyJWT.js"
 lang="javascript" />

The first line, `require('dotenv/config');`, allows the file to use settings from `.env`.

The rest of the file, the function `verifyJWT`, gets the JWT access token from a cookie in the request, and uses `jose.js` to check the signature is correct for your installation of FusionAuth. If either the token is missing or not signed correctly, an error is returned. Otherwise, the next function in the {frontmatter.technology} middleware stack is called.

Now you'll add one more file to check if the user has a role with permissions necessary to call an API endpoint.

In the `services` folder, add the file `hasRole.js`, and copy in the following code.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-javascript-express-api/main/complete-application/services/hasRole.js"
 lang="javascript" />

> This function is given a list of roles, any of which allow the user's request. It returns a middleware function that checks if the JWT contains one of the required roles, and either returns an error or allows the next function in the middleware stack to proceed. Note that you do not need to check for the existence of a token in a cookie here because the previous middleware function you wrote already did that.

Reviewer: Also, the note should be an aside.

### Create The Routes
All that's left to do is write the two endpoints that will demonstrate your authentication.

Overwrite `routes/index.js` with the following code.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-javascript-express-api/main/complete-application/routes/index.js"
 lang="javascript" />

> The POST `panic` route uses your `hasRole(['teller'])` function to disallow non-tellers.

Reviewer: ```suggestion
The POST `panic` route uses the `hasRole` function with the `'teller'` parameter to disallow non-tellers.
```

> The GET `make-change` route uses `hasRole(['customer', 'teller']),` to allow either customers or tellers to make change. This route checks that the user has supplied a number for `total`, then either returns change as JSON or an error if the total was invalid.

Reviewer: ```suggestion
The GET `make-change` route uses `hasRole` function with the `['customer', 'teller']` array as a parameter to allow either customers or tellers to make change. This route checks that the user has supplied a number for `total`, then either returns change as JSON or an error if the total was invalid.
```

    "tokenExpirationInstant": 1689353059670,
    "user": {
        ...
    }
}
```

### Make The Request

Make a request to `/make-change` with a query parameter `total=1.02`. Use the `token` as the `app.at` cookie.

```shell
curl --cookie 'app.at=<your_token>' 'http://localhost:3000/make-change?total=1.02'
```

Your response should look like this:

```json
{
  "total":1.02,
  "nickels":20,
  "pennies":2
}
```

You were successfully authorized. You can try making the request without the JWT or with a different string rather than a valid token, and see that you are denied access.

Next, call the `/panic` endpoint because you are in trouble!

```shell
curl --location --request POST 'http://localhost:3000/panic' --cookie 'app.at=<your_token>'
```

This is a POST not a GET because you want all your emergency calls to be non-idempotent.

Your response should look like this:

```shell

> We've called the police!

Reviewer: let's return json here as well

Reviewer 2: Fixed in this PR https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-api/pull/1

```

Nice, help is on the way!

> Now let's try as `customer@example.com`, which has the role `customer`. Acquire a token for `customer@example.com`.

Reviewer: ```suggestion
Now, let's try as `customer@example.com`, which has the role `customer`. Acquire a token for `customer@example.com`.
```

```shell
curl --location 'http://localhost:9011/api/login' \
--header 'Authorization: this_really_should_be_a_long_random_alphanumeric_value_but_this_still_works' \
--header 'Content-Type: application/json' \
--data-raw '{
  "loginId": "customer@example.com",
  "password": "password",
  "applicationId": "e9fdb985-9173-4e01-9d73-ac2d60d1dc8e"
}'
```

> Now use the token in that response to call `/make-change` with a query parameter `total=3.24`.

Reviewer: ```suggestion
Now, use the token in that response to call `/make-change` with a query parameter `total=3.24`.
```

```shell
curl --cookie 'app.at=<your_token>' 'http://localhost:3000/make-change?total=3.24'
```

Your response should look like this:

```json
{
    "total": 3.24,
    "nickels": 64,
    "pennies": 4
}
```

> So far so good. Now let's try to call the `/panic` endpoint. (We're adding the `-i` flag to see the headers of the response.)

Reviewer: ```suggestion
So far so good. Now, let's try to call the `/panic` endpoint. (We're adding the `-i` flag to see the headers of the response.)
```

</Aside>

FusionAuth will be initially configured with these settings:

* Your client Id is: `e9fdb985-9173-4e01-9d73-ac2d60d1dc8e`.
* Your client secret is: `super-secret-secret-that-should-be-regenerated-for-production`.
* Your example teller username is `teller@example.com` and your password is `password`. They have the role `teller`.
* Your example customer username is `customer@example.com` and your password is `password`. They have the role `customer`.
* Your admin username is `admin@example.com` and your password is `password`.
* Your fusionAuthBaseUrl is 'http://localhost:9011/'.

You can log in to the [FusionAuth admin UI](http://localhost:9011/admin) and look around if you want, but with Docker/Kickstart everything will already be configured correctly.

<Aside type="caution">
  The `.env` and `kickstart.json` files contain passwords. In a real application, always add these files to your `.gitignore` file and never commit secrets to version control.
</Aside>

### Create A Basic {frontmatter.technology} Application

> Now you are going to create an {frontmatter.technology} API application. While this section builds a simple API, you can use the same configuration to integrate an existing API with FusionAuth.

Reviewer: ```suggestion
Now, you are going to create an {frontmatter.technology} API application. While this section builds a simple API, you can use the same configuration to integrate an existing API with FusionAuth.
```

- `make-change`: This endpoint will allow you to call GET with a `total` amount and receive a response indicating how many nickels and pennies are needed to make change. Valid roles are `customer` and `teller`.
- `panic`: Tellers may call this POST endpoint to call the police in case of an incident. The only valid role is `teller`.

Both endpoints will be protected such that a valid JSON web token (JWT) will be required in a cookie or Authorization header in order for the endpoints to be accessed. Additionally, the JWT must have a `roles` claim containing the appropriate role to use the endpoint.

Create the starter app called `your-application` for this API.

```shell
npx express-generator@latest your-application --no-view
cd your-application
npm install
npm audit fix --force
npm install dotenv
npm install jose
```

Here you set up your application and add project dependencies. The [`dotenv`](https://www.npmjs.com/package/dotenv) library allows the code to read your `.env` environment variable file. Additionally, the [`jose`](https://www.npmjs.com/package/jose) library allows you to validate the authenticity of the JWT.

> Now create the `.env` file in your `your-application` folder and add the following to it.

Reviewer: ```suggestion
Now, create the `.env` file in your `your-application` folder and add the following to it.
```

Overwrite `routes/index.js` with the following code.

<RemoteCode url="https://raw.githubusercontent.com/FusionAuth/fusionauth-quickstart-javascript-express-api/main/complete-application/routes/index.js"
 lang="javascript" />

The POST `panic` route uses the `hasRole` function with the `'teller'` parameter to disallow non-tellers.

The GET `make-change` route uses `hasRole` function with the `['customer', 'teller']` array as a parameter to allow either customers or tellers to make change. This route checks that the user has supplied a number for `total`, then either returns change as JSON or an error if the total was invalid.

## Run The Application

Start the API server by running the command below in the `your-application` directory:

```shell
npm run start
```

### Get A Token

There are [several ways to acquire a token](/docs/v1/tech/oauth/) in FusionAuth, but for this example, you will use the [Login API](/docs/v1/tech/apis/login) to keep things simple.

First, let's try the requests as the `teller@example.com` user. Based on the configuration, this user has the `teller` role and should be able to use both `/make-change` and `/panic`.

> Acquire an access token for `teller@example.com` by making the following request.

Reviewer: ```suggestion
Acquire an access token for `teller@example.com` by making the following request in a new terminal window.
```

</Aside>

Make a request to `/make-change` with a query parameter `total=1.02`. Use the `token` as the `app.at` cookie.

```shell
curl --cookie 'app.at=<your_token>' 'http://localhost:3000/make-change?total=1.02'
```

Alternatively you can make the same request by passing your token in the `Authorization` header.

```shell
curl --location 'http://localhost:3000/make-change?total=1.02' \
--header 'Authorization: Bearer <your_token>'
```

Your response should look like this:

```json
{
  "total":1.02,
  "nickels":20,
  "pennies":2
}
```

You were successfully authorized. You can try making the request without the JWT or with a different string rather than a valid token, and see that you are denied access.

Next, call the `/panic` endpoint because you are in trouble!

```shell
curl --location --request POST 'http://localhost:3000/panic' --cookie 'app.at=<your_token>'
```

This is a POST not a GET because you want all your emergency calls to be non-idempotent.

Your response should look like this:

```json

> {

Reviewer: my response did not contain json, just the message.

Reviewer 2: @mark-robustelli the change which returns json is in this open PR https://github.com/FusionAuth/fusionauth-quickstart-javascript-express-api/pull/1. Did you use that code when testing.

### Make The Request

<Aside type="note">

> The code is set up to extract the token from either a cookie or the Authorization header so depending on your preference you can replace `--cookie 'app.at=<your_token>'` with `--header 'Authorization: Bearer <your_token>'` when making requests to the API.

Reviewer 3: ```suggestion
The code is set up to extract the token from either a cookie or the Authorization header so depending on your preference you can replace `--cookie 'app.at=<your_token>'` with `--header 'Authorization: Bearer <your_token>'` when making requests to the API.

If you use a cookie, make sure you store it in a secure, HTTPOnly cookie to avoid exfiltration attacks. See [Storing OAuth Tokens](/articles/oauth/oauth-token-storage) for more information.
```
